runtime/src/lib.rs

```diff
@@ -158,6 +166,87 @@ pub fn native_version() -> NativeVersion {
 
 const NORMAL_DISPATCH_RATIO: Perbill = Perbill::from_percent(75);
 
+impl ChainExtension<Runtime> for ExampleExtension
+where
+	Runtime: SysConfig + pallet_contracts::Config,
+	<Runtime as SysConfig>::AccountId: UncheckedFrom<<Runtime as SysConfig>::Hash> + AsRef<[u8]>,
+{
+	fn call<E: Ext>(func_id: u32, env: Environment<E, InitState>) -> Result<RetVal, DispatchError>
+	where
+		E: Ext<T = Runtime>,
+		<E::T as SysConfig>::AccountId: UncheckedFrom<<E::T as SysConfig>::Hash> + AsRef<[u8]>,
+	{
+		// Match on function id assigned in the contract
+		match func_id {
+			// do_store_in_runtime
+			1 => {
+				use pallet_template::WeightInfo;
+				let mut env = env.buf_in_buf_out();
+				// retrieve argument that was passed in smart contract invocation
+				let value: u32 = env.read_as()?;
+				// Capture weight for the main action being performed by the extrinsic
+				let base_weight: Weight =
+					<Runtime as pallet_template::Config>::WeightInfo::insert_number(value);
+				// Add some weight for the contract invocation
+				let overhead = <Runtime as pallet_contracts::Config>::Schedule::get()
+					.host_fn_weights
+					.debug_message;
+				env.charge_weight(base_weight.saturating_add(overhead))?;
+				let caller = env.ext().caller().clone();
+
+				crate::pallet_template::Pallet::<Runtime>::insert_number(
+					RawOrigin::Signed(caller).into(),
+					value,
+				)?;
+			},
+			// do_balance_transfer
+			2 => {
+				let mut env = env.buf_in_buf_out();
+				// Retrieve arguments
+				let (transfer_amount, recipient): (u32, AccountId) = env.read_as()?;
+				let base_weight = <Runtime as pallet_contracts::Config>::Schedule::get()
+					.host_fn_weights
+					.call_transfer_surcharge;
+				let overhead = <Runtime as pallet_contracts::Config>::Schedule::get()
+					.host_fn_weights
+					.debug_message;
+				env.charge_weight(base_weight.saturating_add(overhead))?;
+
+				let caller = env.ext().caller().clone();
+				let recipient_account = sp_runtime::MultiAddress::Id(recipient);
+
+				pallet_balances::Pallet::<Runtime>::transfer(
+					RawOrigin::Signed(caller).into(),
+					recipient_account,
+					transfer_amount.into(),
+				)
+				.map_err(|d| d.error)?;
+			},
+			// do_get_balance
+			3 => {
+				let mut env = env.buf_in_buf_out();
+				// Retrieve argument
+				let account: AccountId = env.read_as()?;
+				let result = pallet_balances::Pallet::<Runtime>::free_balance(account).encode();
+				env.write(&result, false, None)
+					.map_err(|_| "Encountered an error when querying balance.")?;
+			},
+			// do_get_from_runtime
+			4 => {
+				let mut env = env.buf_in_buf_out();
+				let result = TemplateModule::get_value().encode();
+				env.write(&result, false, None)
+					.map_err(|_| "Encountered an error when retrieving runtime storage value.")?;
+			},
```

Comment:
Note that these don't charge any weight. I'm unsure whether that's correct.

Reply:
Discussed in DM and committed changes with these weights